wini2c/i2capi.cpp

```diff
@@ -22,126 +24,194 @@ PyErr_SetString(PyExc_TypeError, errmsg); \
+        return nullptr;
+    }
+
+    return i2cdevice;
+}
+
+static I2cDevice^ GetDeviceLightning(int id, I2cConnectionSettings^ settings) {
+    auto asyncop = I2cController::GetControllersAsync(LightningI2cProvider::GetI2cProvider());
+    while (asyncop->Status != AsyncStatus::Completed) {
+        if (asyncop->Status == AsyncStatus::Error) {
+            PyErr_Format(PyExc_RuntimeError, "Could not find i2c controller: %d", asyncop->ErrorCode);
+            return nullptr;
+        }
+        Sleep(50);
+    }
+
+    auto controllers = asyncop->GetResults();
+    if (controllers == nullptr) {
+        PyErr_Format(PyExc_RuntimeError, "Could not find i2c controller");
+        return nullptr;
+    }
+
+    if (id < 0 || static_cast<unsigned int>(id) >= controllers->Size) {
+        PyErr_Format(PyExc_RuntimeError, "Could not find i2c controller id: '%d'", id);
+        return nullptr;
+    }
+
+    auto controller = controllers->GetAt(id);
+    auto i2cdevice = controller->GetDevice(settings);
+    if (i2cdevice == nullptr) {
+        PyErr_SetString(PyExc_RuntimeError, "Could not find I2C device specified");
+        return nullptr;
+    }
+
+    return i2cdevice;
+}
 
-			auto i2cop = I2cDevice::FromIdAsync(id, settings);
-			while (i2cop->Status != AsyncStatus::Completed) {
-				if (i2cop->Status == AsyncStatus::Error) {
-					PyErr_Format(PyExc_RuntimeError, "Could get I2C device: %d", i2cop->ErrorCode);
-					return NULL;
-				}
-				Sleep(50);
-			}
-
-			auto i2cdevice = i2cop->GetResults();
-            if (i2cdevice == nullptr) {
-                PyErr_SetString(PyExc_RuntimeError, "Could not find I2C device specified");
-                return NULL;
+static I2cDevice^ GetDevice(int id, I2cConnectionSettings^ settings) {
+    if (LightningProvider::IsLightningEnabled) {
+        return GetDeviceLightning(id, settings);
+    } else {
+        return GetDeviceInboxDriver(id, settings);
+    }
+}
+
+static I2cConnectionSettings^ GetI2cConnectionSettings(int slaveAddress, int busSpeed, int sharingMode) {
+    I2cConnectionSettings^ settings = ref new I2cConnectionSettings(slaveAddress);
+
+    switch (busSpeed)
+    {
+    case FASTSPEED:
+        settings->BusSpeed = I2cBusSpeed::FastMode;
+        break;
+    case STANDARDSPEED:
+        settings->BusSpeed = I2cBusSpeed::StandardMode;
+        break;
+    default:
+        PyErr_Format(PyExc_TypeError, "Invalid I2C bus speed specified '%d'", busSpeed);
+        return nullptr;
+    }
+
+    switch (sharingMode)
+    {
+    case SHAREDMODE:
+        settings->SharingMode = I2cSharingMode::Shared;
+        break;
+    case EXCLUSIVEMODE:
+        settings->SharingMode = I2cSharingMode::Exclusive;
+        break;
+    default:
+        PyErr_Format(PyExc_TypeError, "Invalid I2C sharing mode specified '%d'", sharingMode);
+        return nullptr;
+    }
+
+    return settings;
+}
+
+extern "C" {
+    void enable_lightning_if_available() {
```

Comment:
Not sure what you mean.  We need to call this set LowLevelDevicesController.DefaultProvider before getting any controllers or devices.